tests/johnsnowlabs/test_jsl_api.py

```diff
@@ -0,0 +1,782 @@
+    d1 = d1.drop(columns=[c for c in d1.columns if c not in d2.columns or 'confidence' in c])
+    d2 = d2.drop(columns=[c for c in d2.columns if c not in d1.columns or 'confidence' in c])
+
+    def lower_strings(df):
+        for c in df.columns:
+            try:
+                df[c] = df[c].str.lower()
+            except:
+                pass
+        return df
+
+    d1 = lower_strings(d1)
+    d2 = lower_strings(d2)
+    return d1, d2
+
+
+def validate_model(original_model, new_model):
+    d1, d2 = original_model.predict("Hello World"), new_model.predict("Hello World")
+    if isinstance(d2, str):
+        d2 = pd.DataFrame(json.loads(d2)).drop(columns=['index']).reset_index().drop(columns=['index'])
+    else:
+        d2 = d2.reset_index().drop(columns=['index'])
+    d1 = d1.reset_index().drop(columns=['index'])
+
+    # TODO fix: column names may change before/after save and Confidences change
+    d1 = d1.drop(columns=[c for c in d1.columns if 'confidence' in c])
+    d2 = d2.drop(columns=[c for c in d2.columns if 'confidence' in c])
+    d1.columns = [f'c_{i}' for i in range(len(d1.columns))]
+    d2.columns = [f'c_{i}' for i in range(len(d2.columns))]
+    assert d1.equals(d2)
+
+
+@pytest.fixture
+def jsl_model(model_path):
+    model = load_and_init_model()
+    yield model
+    # shutil.rmtree(model_path)
+
+
+@pytest.fixture
+def model_path(tmpdir):
+    return os.path.join(str(tmpdir), "model")
+
+
+@pytest.fixture
+def spark_custom_env(tmpdir):
+    conda_env = os.path.join(str(tmpdir), "conda_env.yml")
+    additional_pip_deps = ["pyspark", "pytest"]
+    if Version(pyspark.__version__) <= Version("3.3.2"):
+        # Versions of PySpark <= 3.3.2 are incompatible with pandas >= 2
+        additional_pip_deps.append("pandas<2")
+    _mlflow_conda_env(conda_env, additional_pip_deps=additional_pip_deps)
+    return conda_env
+
+
+def score_model_as_udf(model_uri, result_type='string'):
+    spark = mlflow.johnsnowlabs.get_or_create_sparksession()
+    pandas_df = pd.DataFrame({'text': ['Hello World']})
+    spark_df = spark.createDataFrame(pandas_df).coalesce(1)
+    pyfunc_udf = spark_udf(
+        spark=spark, model_uri=model_uri, env_manager="virtualenv", result_type=result_type,
+    )
+    new_df = spark_df.withColumn("prediction", pyfunc_udf(*pandas_df.columns))
+    return [x["prediction"] for x in new_df.collect()]
+
+
+def test_hadoop_filesystem(tmpdir, jsl_model):
+    # we need an active spark session, so we just pass jsl_model to init a session
+    # copy local dir to and back from HadoopFS and make sure the results match
+    from mlflow.spark import _HadoopFileSystem as FS
+    test_dir_0 = os.path.join(str(tmpdir), "expected")
+    test_file_0 = os.path.join(test_dir_0, "root", "file_0")
+    test_dir_1 = os.path.join(test_dir_0, "root", "subdir")
+    test_file_1 = os.path.join(test_dir_1, "file_1")
+    os.makedirs(os.path.dirname(test_file_0))
+    with open(test_file_0, "w") as f:
```

Comment:
Can this be changed to pathlib conventions so that we can remove all of this boilerplate?

i.e., 
``` python
test_file_0 = tmpdir / "root" / "file_0"
test_file_0.write_text("test0")
```

Reply:
removed this test